p2p/server.go

```diff
@@ -599,12 +619,32 @@ running:
 		case n := <-srv.removestatic:
 			// This channel is used by RemovePeer to send a
 			// disconnect request to a peer and begin the
-			// stop keeping the node connected
+			// stop keeping the node connected.
 			srv.log.Debug("Removing static node", "node", n)
 			dialstate.removeStatic(n)
 			if p, ok := peers[n.ID]; ok {
 				p.Disconnect(DiscRequested)
 			}
+		case n := <-srv.addtrusted:
+			// This channel is used by AddTrustedPeer to add an enode
+			// to the trusted node set.
+			srv.log.Debug("Adding trusted node", "node", n)
+			trusted[n.ID] = true
+			// Mark any already-connected peer as trusted
+			if p, ok := peers[n.ID]; ok {
+				p.rw.flags |= trustedConn
```

Comment:
Since this can happen any time during the peer's lifecycle, can't this write collide with a read in func (p *Peer) Info() ? I am not very familiar with this part but it looks like we assumed until now that flags don't change after the peer is connected. @fjl opinion? Should we not use locks or atomic read/write if we change the flags of live peers?

Reply:
That's a good question. `go test -race ./p2p` didn't catch it but you might be right.

I'm happy to bolt the flags down with a mutex or atomic. Let me know what you'd prefer. :)

Reply:
I think it's OK to just add the ID to the trusted set without modifying flags of existing connections. The flag is not used after handshake anyway.

Reply:
@fjl Hm, I worry that has potential for breaking other code that relies on the peer state. Such as if you do `admin.peers`, it will return incorrect metadata, right?

Reply:
We would need a lock then and I want to avoid that ;)

Reply:
@fjl Just rebased and pushed another commit with more testing for this, still passes with `-race` on. I'm fairly convinced that these variables are only ever written and read inside the server event loop, so I don't think there is a race condition.

If you still prefer, I can get rid of updating the flags. We might be able to get rid of the flags altogether and just read the state from the server lookup maps on-demand, but that may be a somewhat larger refactor.

Reply:
It passes because nothing calls p.Inbound in those tests. That would race. Maybe the best way to avoid a race is to just cache inbound flag in peer as a bool.

Sorry for being so pedantic about this, the PR is good otherwise.

Reply:
@fjl Nah that's not pedantic; I don't want to introduce a race, but I was being fixated on the parts that I used.

Let me make some more tweaks, will ping again.